Домашнее задание

RushianHaker opened the pull request:
**Урок**: Введение в Spring Framework - домашняя работа
**Цель**: Создать приложение с помощью Spring IoC, чтобы познакомиться с основной функциональностью IoC, на которой строится весь Spring.
**Результат**: простое приложение, сконфигурированное XML-контекстом.

helloDearSupportTeam committed:
инициализация проекта

helloDearSupportTeam committed:
закончил выполнение первого задания

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/service/impl/TestServiceImpl.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    public HashMap<Question, ArrayList<Answer>> getTestList() {
+        var mapQuestions = new HashMap<Question, ArrayList<Answer>>();
```
С т.з. ООП будет лучше, если ответы будут частью сущности вопроса

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/service/impl/TestServiceImpl.java:
```diff
@@ -0,0 +1,48 @@
+    public HashMap<Question, ArrayList<Answer>> getTestList() {
+        var mapQuestions = new HashMap<Question, ArrayList<Answer>>();
+
+        try (var reader = new CSVReader(new InputStreamReader(new ClassPathResource("/test.csv").getInputStream()))) {
+            String[] nextLine;
+
+            while ((nextLine = reader.readNext()) != null) {
+                var listAnswers = new ArrayList<Answer>();
+
+                listAnswers.add(answerDao.createAnswer(nextLine[1]));
+                listAnswers.add(answerDao.createAnswer(nextLine[2]));
+
+                mapQuestions.put(questionDao.createQuestion(nextLine[0]), listAnswers);
+            }
+        } catch (CsvValidationException | IOException e) {
+            throw new RuntimeException("Csv read error: " + e);
+        }
```
Не совсем. Скрыть источник данных, как раз задача дао. Т.е. мы должны в дао вызвать условный `findAll` и получить список вопросов. Сохранения же вопросов, в нашем приложении нет. Мы их только грузим

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/service/impl/TestServiceImpl.java:
```diff
@@ -0,0 +1,48 @@
+    public HashMap<Question, ArrayList<Answer>> getTestList() {
+        var mapQuestions = new HashMap<Question, ArrayList<Answer>>();
+
+        try (var reader = new CSVReader(new InputStreamReader(new ClassPathResource("/test.csv").getInputStream()))) {
```
Из ДЗ
>4. Имя ресурса с вопросами (CSV-файла) необходимо захардкодить строчкой в XML-файле с контекстом.

у нас же оно в коде

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/service/impl/TestServiceImpl.java:
```diff
@@ -0,0 +1,48 @@
+        this.answerDao = answerDao;
+    }
+
+    public HashMap<Question, ArrayList<Answer>> getTestList() {
```
Слева всегда лучше писать абстракцию, а не реализацию. В данном случае это `Map` и `List`. Ну и мапы крайне плохо воспринимающийся тип данных. Поэтому его лучше по возможности не использовать в публичных методах

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/utils/PrintUtil.java:
```diff
@@ -0,0 +1,8 @@
+package ru.otus.testing.utils;
+
+
+public final class PrintUtil {
```
👍 

Только я бы не использовал слово `Util`. Лучше `Service`

stvort commented on spring-01/testing-students/src/test/java/ru/otus/testing/TestingStudentsApplicationTests.java:
```diff
@@ -0,0 +1,9 @@
+
+import org.junit.jupiter.api.Test;
+
+class TestingStudentsApplicationTests {
+    @Test
+    void contextLoads() {
+    }
+}
```
Такие тесты обычно создаются при генерации SpringBoot проекта. Только там еще `@SpringBootTest` висит. И из-за него поднимается тестовый контекст. И если не поднимется, то упадет любой тест внутри класса. Поэтому даже пустой тест проверяет то, что контекст может подняться. В данном случае это просто пустой тест, который ничего не проверяет. Ну и даже если бы у нас был `@SpringBootTest`, то это был бы интеграционный тест. А по заданию нужно сделать юнит-тест

helloDearSupportTeam committed:
нейминг

helloDearSupportTeam committed:
Имя ресурса с вопросами (CSV-файла) необходимо захардкодить строчкой в XML-файле с контекстом.

helloDearSupportTeam committed:
(С т.з. ООП будет лучше, если ответы будут частью сущности вопроса) И (использую абстракцию List - в публичном методе)

helloDearSupportTeam committed:
перенес в другой пакет сервис

helloDearSupportTeam committed:
Не совсем. Скрыть источник данных, как раз задача дао. Т.е. мы должны в дао вызвать условный findAll и получить список вопросов. Сохранения же вопросов, в нашем приложении нет. Мы их только грузим

helloDearSupportTeam committed:
добавил аннотацию для тестирования

helloDearSupportTeam committed:
аннотация для правильной сборки

helloDearSupportTeam committed:
новый тест сервиса, проверяем работу метода getTestList()

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/model/Answer.java:
```diff
@@ -0,0 +1,17 @@
+package ru.otus.testing.model;
+
+public class Answer {
+    private String answer;
```
Ответ может быть правильным или не правильным еще

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/dao/AnswerDao.java:
```diff
@@ -0,0 +1,7 @@
+
+import ru.otus.testing.model.Answer;
+
+public interface AnswerDao {
```
Нам не нужен этот дао

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/dao/impl/QuestionDaoImpl.java:
```diff
@@ -0,0 +1,50 @@
+            while ((nextLine = reader.readNext()) != null) {
+                var listAnswers = new ArrayList<Answer>();
+
+                listAnswers.add(answerDao.createAnswer(nextLine[1]));
+                listAnswers.add(answerDao.createAnswer(nextLine[2]));
```
В роли источника данных для вопросов и ответов выступает csv-файл. Эти строки в концепции дао значили бы, что мы при чтении вопросов из источника данных создаем в нем варианты ответов, что очевидно не так

RushianHaker replied:
убрал лишнее DAO

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/dao/impl/QuestionDaoImpl.java:
```diff
@@ -0,0 +1,50 @@
+                listAnswers.add(answerDao.createAnswer(nextLine[1]));
+                listAnswers.add(answerDao.createAnswer(nextLine[2]));
+
+                listQuestions.add(createQuestion(nextLine[0], listAnswers));
```
А тут соответственно создаем вопросы в источнике данных при чтении их из него

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/dao/impl/QuestionDaoImpl.java:
```diff
@@ -0,0 +1,50 @@
+            while ((nextLine = reader.readNext()) != null) {
+                var listAnswers = new ArrayList<Answer>();
+
+                listAnswers.add(answerDao.createAnswer(nextLine[1]));
+                listAnswers.add(answerDao.createAnswer(nextLine[2]));
```
Я бы не ограничивал количество возможных вариантов в коде, а прочитал все что положили

stvort replied:
При отсутствии в `nextLine` нужного количества элементов, тут будет исключение, которое мы не ловим

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/service/impl/TestServiceImpl.java:
```diff
@@ -0,0 +1,20 @@
+        this.questionDao = questionDao;
+    }
+
+    public List<Question> getTestList() {
+        return questionDao.findAll();
+    }
```
`getTestList` возвращает не список тестов, как можно было бы подумать из названия

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/service/utils/PrintService.java:
```diff
@@ -0,0 +1,8 @@
+
+
+public final class PrintService {
+    public static void print(String text) {
```
Не нужно использовать статические методы для бизнес-логики. Протестировать классы, где такое используется с помощью юнит-тестов будет тяжело. Мы обсуждали этот момент на занятии

RushianHaker replied:
вынес СЛУЖЕБНЫЙ FINAL класс из пакета с бизнес логикой

stvort commented on spring-01/testing-students/src/test/java/ru/otus/testing/TestingStudentsApplicationTests.java:
```diff
@@ -0,0 +1,12 @@
+import org.junit.runner.RunWith;
+import org.springframework.test.context.junit4.SpringRunner;
+
+@RunWith(SpringRunner.class)
```
Это от JUnit4, лучше использовать пятый. Ну и JUnit4 нет в зависимостях, программа не компилируется

stvort replied:
Опять же. Тесты с контекстом, это интеграционные тесты. А по заданию надо написать юнит-тест

Commit:
При отсутствии в nextLine нужного количества элементов, тут будет исключение, которое мы не ловим
 и
Я бы не ограничивал количество возможных вариантов в коде, а прочитал все что положили

Commit:
нейминг метода

Commit:
удалил интеграционный тест

Commit:
поправил зависимости

Commit:
исправил поломанный тест

Commit:
вынес пакет с util (служебным) классом, который не относиться к бизнес логике

Commit:
убрал лишнее DAO

Commit:
А тут соответственно создаем вопросы в источнике данных при чтении их из него

Commit:
добавил индикатор корректности

Commit:
обновил тест новым полем корректности ответа

Commit:
упростил тест

helloDearSupportTeam committed:
checkStyle

helloDearSupportTeam committed:
вынес получение вопросов в отдельный метод - checkStyle

helloDearSupportTeam committed:
чистка мусора

helloDearSupportTeam committed:
добавил зависимость для сборки jar с зависимостями

helloDearSupportTeam committed:
чистка

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/dao/impl/QuestionDaoImpl.java:
```diff
@@ -0,0 +1,52 @@
+            String[] nextLine;
+
+            while ((nextLine = reader.readNext()) != null) {
+                if (nextLine.length > 2) {
```
А если меньше? Это же исключительная ситуация?

RushianHaker replied:
Меньше быть не может, так как по условию: есть вопрос (это 1) и есть как минимум 1 ответ (это 2), если нет ни вопроса, ни ответа, то он не будет это обрабатывать, а также если есть вопрос, но нет ответа,  то он просто выведет вопрос.
Какое условие вы подразумеваете нужно здесь обработать ?

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/dao/impl/QuestionDaoImpl.java:
```diff
@@ -0,0 +1,52 @@
+            }
+
+        } catch (CsvValidationException | IOException e) {
+            throw new RuntimeException("Csv read error: " + e);
```
Так мы подменяем исключение, а лучше завернуть.
```
throw new RuntimeException("Csv read error", e)
```

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/utils/PrintService.java:
```diff
@@ -0,0 +1,8 @@
+package ru.otus.testing.utils;
+
+
+public final class PrintService {
```
Этот сервис будет использоваться в качестве зависимости для вывода всех сообщений. Так что ему бы интерфейс, потом сделать сервис полноценным бином и статический метод сделать полноценным методом экземпляра.

>вынес СЛУЖЕБНЫЙ FINAL класс из пакета с бизнес логикой

Не совсем. Этот служебный класс все еще используется в бизнес логике. И все еще мешает ее тестировать без себя. Ну и при смене типа вывода, придется везде пройтись и заменить вызовы такого кода на что-то другое. Использование такого рода классов со статическими методами нарушает, как минимум два принципа SOLID, SRP + DI, OCP тоже можно притянуть. Это не значит, что хелперы не нужно использовать совсем, `isNull`/`nonNull` и подобные вещи отлично ложатся на эту схему, но они не являются компонентами приложения, а вот класс осуществляющий ввод-ввод - является

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/TestingStudentsApplication.java:
```diff
@@ -0,0 +1,24 @@
+    public static void main(String[] args) {
+        ClassPathXmlApplicationContext context = new ClassPathXmlApplicationContext("/spring-context.xml");
+        TestService service = context.getBean(TestService.class);
+        var listQuestions = service.getTest();
+
+        for (var question : listQuestions) {
+            PrintService.print("\n" + "Question: " + question.getQuestion());
+            PrintService.print("Answers:");
+            for (var answer : question.getAnswer()) {
+                PrintService.print("* " + answer.getAnswer());
+            }
+        }
```
Как мы обсуждали на занятии, никакой логики в `main` быть не должно. Только создание контекста, получение из него одного бина и вызов одного его метода

stvort commented on spring-01/testing-students/src/test/java/ru/otus/testing/service/impl/TestServiceImplTest.java:
```diff
@@ -0,0 +1,30 @@
+
+class TestServiceImplTest {
+
+    ClassPathXmlApplicationContext context = new ClassPathXmlApplicationContext("/spring-context.xml");
+    TestService service = context.getBean(TestService.class);
```
> Опять же. Тесты с контекстом, это интеграционные тесты. А по заданию надо написать юнит-тест

Не исправлено.

С помощью юнит-теста текущий класс можно протестировать следующим образом:
- Сделать мок `QuestionDao` библиотекой `Mockito`
- Настроить мок на отдачу заданного списка вопросов
- Создать `TestService`, передав мок ему в конструктор
- Вызвать метод `getTest` и сравнить результат с тем, на который мы настраивали мок

Commit:
Так мы подменяем исключение, а лучше завернуть.

Commit:
переписал интеграционный-тест на юнит-тест !

Commit:
PrintService бы интерфейс, потом сделать сервис полноценным бином и статический метод сделать полноценным методом экземпляра.
и
Только создание контекста, получение из него одного бина и вызов одного его метода

Commit:
докинул интерфейс

Commit:
добавил проверку на IndexOutOfBoundsException

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/service/TestService.java:
```diff
@@ -0,0 +1,12 @@
+import java.util.List;
+
+public interface TestService {
+    List<Question> getTest();
+
+    void printTest();
```
Это нарушение SRP из SOLID и YAGNY. Для данной работы, лучше чтобы тут было только `printTest`. Если бы нам было нужно и отдавать тест и печатать, то скорее всего было бы два сервиса. Как я понимаю, `getTest()` нужен для того, чтобы протестировать сервис, но лучше для этого дополнительных методов не писать. Результатом метода печати является вывод данных пользователю, а значит надо именно это и проверить. Например, через проверку аргументов переданных в метод мока `PrintService`

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/dao/impl/QuestionDaoImpl.java:
```diff
@@ -0,0 +1,54 @@
+            while ((nextLine = reader.readNext()) != null) {
+                listQuestions.add(new Question(nextLine[0], findAnswers(nextLine)));
+            }
+        } catch (IndexOutOfBoundsException e) {
+            throw new ArrayIndexOutOfBoundsException("Array was throw exception: " + e);
+        } catch (CsvValidationException e) {
+            throw new CsvRuntimeException("User-defined csv validator fails: " + e);
+        } catch (IOException e) {
+            throw new IllegalArgumentException("Reads resource or next line from the buffer was throw exception: " + e);
+        }
```
Еще раз. Не подменяем, а заворачиваем. Желательно в свое бизнес-исключение. Указав оригинальное в качестве `cause`

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/dao/impl/QuestionDaoImpl.java:
```diff
@@ -0,0 +1,54 @@
+        return listQuestions;
+    }
+
+    public List<Answer> findAnswers(String[] nextLine) {
+        var listAnswers = new ArrayList<Answer>();
+        for (int i = 1; i < nextLine.length; i++) {
+            if (nextLine[i].contains("correct:")) {
+                listAnswers.add(new Answer(nextLine[i].replaceFirst("correct:", ""), true));
+            } else {
+                listAnswers.add(new Answer(nextLine[i], false));
+            }
+        }
+        return listAnswers;
+    }
```
Зачем это `public`?

stvort commented on spring-01/testing-students/src/test/java/ru/otus/testing/service/impl/TestServiceImplTest.java:
```diff
@@ -0,0 +1,42 @@
+
+
+class TestServiceImplTest {
+    QuestionDao questionDao = mock(QuestionDao.class);
+    TestService service = new TestServiceImpl(questionDao, new PrintServiceImpl());
```
`private`

stvort commented on spring-01/testing-students/src/test/java/ru/otus/testing/service/impl/TestServiceImplTest.java:
```diff
@@ -0,0 +1,42 @@
+
+
+class TestServiceImplTest {
+    QuestionDao questionDao = mock(QuestionDao.class);
+    TestService service = new TestServiceImpl(questionDao, new PrintServiceImpl());
```
В юинт-тестах все зависимости мокаются

RushianHaker replied:
спасибо за объяснение такого нюанса

stvort commented on spring-01/testing-students/src/test/java/ru/otus/testing/service/impl/TestServiceImplTest.java:
```diff
@@ -0,0 +1,42 @@
+import static org.mockito.Mockito.when;
+
+
+class TestServiceImplTest {
```
```
class TestServiceImplTest {
    private QuestionDao questionDao;
    private PrintService printService;
    private TestService service;

    @BeforeEach
    void setUp() {
        questionDao = mock(QuestionDao.class);
        printService = mock(PrintService.class);
        service = new TestServiceImpl(questionDao, printService);
    }

    @Test
    void getTestListTest() {
        var answer1 = new Answer("test answer 1", true);
        var answer2 = new Answer("test answer 2", false);
        var question = new Question("test question", List.of(answer1, answer2));

        when(questionDao.findAll()).thenReturn(List.of(question));

        service.printTest();
        ArgumentCaptor<String> captor = ArgumentCaptor.forClass(String.class);
        verify(printService, times(4)).print(captor.capture());

        String actualOutput = captor.getAllValues().stream()
                .collect(Collectors.joining(System.lineSeparator()));
        assertTrue(actualOutput.contains(question.getQuestion()));
        assertTrue(actualOutput.contains(answer1.getAnswer()));
        assertTrue(actualOutput.contains(answer2.getAnswer()));
    }
}
```

helloDearSupportTeam committed:
private

helloDearSupportTeam committed:
создал свое исключение

helloDearSupportTeam committed:
обернул в свое бизнес-исключение

helloDearSupportTeam committed:
тест

helloDearSupportTeam committed:
SRP из SOLID и YAGNY

helloDearSupportTeam committed:
нейминг

stvort commented on spring-01/testing-students/src/main/java/ru/otus/testing/exception/QuestionDaoImplException.java:
```diff
@@ -0,0 +1,8 @@
+package ru.otus.testing.exception;
+
+
+public class QuestionDaoImplException extends RuntimeException {
```
Просто `QuestionDaoException`. Нам не нужно писать для каждой реализации дао свое исключение. Они все могут кидать одно и то же

helloDearSupportTeam committed:
нейминг
